File: src/Optimization/StructureAware/HierarchicalGenome.cs
using Roguelike.Data;
using Roguelike.Core.Map;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Roguelike.Optimization
{
    /// <summary>
    /// A hierarchical genome that reduces the parameter space from ~500 to ~50 parameters
    /// by using multiplicative layers and only overriding specific outliers.
    /// </summary>
    public class HierarchicalGenome
    {
        // TOP LEVEL: Global Multipliers (5 params)
        public float GlobalDamageMultiplier { get; set; } = 1.0f;
        public float GlobalHealthMultiplier { get; set; } = 1.0f;
        public float GlobalBlockMultiplier { get; set; } = 1.0f;
        public float GlobalManaCostMultiplier { get; set; } = 1.0f;
        public float GlobalGoldMultiplier { get; set; } = 1.0f;

        // MID LEVEL: Progression Scaling (9 params)
        public float EarlyGameDamageScaling { get; set; } = 1.0f;   // Floors 1-5
        public float MidGameDamageScaling { get; set; } = 1.0f;     // Floors 6-10
        public float LateGameDamageScaling { get; set; } = 1.0f;    // Floors 11-15
        
        public float EarlyGameHealthScaling { get; set; } = 1.0f;
        public float MidGameHealthScaling { get; set; } = 1.0f;
        public float LateGameHealthScaling { get; set; } = 1.0f;
        
        public float EarlyGameBlockScaling { get; set; } = 1.0f;
        public float MidGameBlockScaling { get; set; } = 1.0f;
        public float LateGameBlockScaling { get; set; } = 1.0f;

        // CATEGORY LEVEL: Type-Based Scaling (15 params)
        public Dictionary<CardType, float> CardTypeScalars { get; set; } = new()
        {
            { CardType.Attack, 1.0f },
            { CardType.Skill, 1.0f },
            { CardType.Power, 1.0f }
        };
        
        public Dictionary<int, float> CardStarScalars { get; set; } = new()
        {
            { 1, 1.0f }, { 2, 1.0f }, { 3, 1.0f }, { 4, 1.0f }, { 5, 1.0f }
        };
        
        public Dictionary<int, float> EnemyStarScalars { get; set; } = new()
        {
            { 1, 1.0f }, { 2, 1.0f }, { 3, 1.0f }, { 4, 1.0f }, { 5, 1.0f }
        };

        // ROOM DISTRIBUTION (8 params)
        public Dictionary<RoomType, float> RoomTypeWeights { get; set; } = new()
        {
            { RoomType.Monster, 45f },
            { RoomType.Elite, 12f },
            { RoomType.Event, 22f },
            { RoomType.Shop, 8f },
            { RoomType.Rest, 13f }
        };
        
        public float MonsterStarRatio { get; set; } = 0.5f;
        public float EliteStarRatio { get; set; } = 0.5f;
        public float RestHealingScalar { get; set; } = 1.0f;

        // HERO BASELINE (3 params)
        public float HeroHealthScalar { get; set; } = 1.0f;
        public float HeroStartGoldScalar { get; set; } = 1.0f;
        public int HeroManaOffset { get; set; } = 0;

        // INDIVIDUAL OVERRIDES (Sparse - only for problem cards/enemies)
        public Dictionary<string, float> CardDamageOverrides { get; set; } = new();
        public Dictionary<string, float> CardManaCostOverrides { get; set; } = new();
        public Dictionary<string, float> EnemyHealthOverrides { get; set; } = new();

        // INITIALIZATION
        public HierarchicalGenome() { }

        public void Randomize(Random rng)
        {
            // Global multipliers: 0.7 - 1.3
            GlobalDamageMultiplier = (float)(0.7 + rng.NextDouble() * 0.6);
            GlobalHealthMultiplier = (float)(0.7 + rng.NextDouble() * 0.6);
            GlobalBlockMultiplier = (float)(0.7 + rng.NextDouble() * 0.6);
            GlobalManaCostMultiplier = (float)(0.85 + rng.NextDouble() * 0.3);
            GlobalGoldMultiplier = (float)(0.7 + rng.NextDouble() * 0.6);

            // Progression scaling: 0.8 - 1.4
            EarlyGameDamageScaling = (float)(0.8 + rng.NextDouble() * 0.3);
            MidGameDamageScaling = (float)(1.0 + rng.NextDouble() * 0.3);
            LateGameDamageScaling = (float)(1.2 + rng.NextDouble() * 0.3);
            
            EarlyGameHealthScaling = (float)(0.8 + rng.NextDouble() * 0.3);
            MidGameHealthScaling = (float)(1.0 + rng.NextDouble() * 0.3);
            LateGameHealthScaling = (float)(1.2 + rng.NextDouble() * 0.3);
            
            EarlyGameBlockScaling = (float)(0.9 + rng.NextDouble() * 0.2);
            MidGameBlockScaling = (float)(1.0 + rng.NextDouble() * 0.2);
            LateGameBlockScaling = (float)(1.0 + rng.NextDouble() * 0.2);

            // Card type scaling: 0.85 - 1.15
            foreach (var key in CardTypeScalars.Keys.ToList())
                CardTypeScalars[key] = (float)(0.85 + rng.NextDouble() * 0.3);

            // Star rating scaling: 0.9 - 1.1
            foreach (var key in CardStarScalars.Keys.ToList())
                CardStarScalars[key] = (float)(0.9 + rng.NextDouble() * 0.2);
            
            foreach (var key in EnemyStarScalars.Keys.ToList())
                EnemyStarScalars[key] = (float)(0.9 + rng.NextDouble() * 0.2);

            // Room weights
            RoomTypeWeights[RoomType.Monster] = (float)(30 + rng.NextDouble() * 30);
            RoomTypeWeights[RoomType.Elite] = (float)(5 + rng.NextDouble() * 20);
            RoomTypeWeights[RoomType.Event] = (float)(10 + rng.NextDouble() * 25);
            RoomTypeWeights[RoomType.Shop] = (float)(5 + rng.NextDouble() * 15);
            RoomTypeWeights[RoomType.Rest] = (float)(5 + rng.NextDouble() * 20);
            
            MonsterStarRatio = (float)(0.3 + rng.NextDouble() * 0.4);
            EliteStarRatio = (float)(0.3 + rng.NextDouble() * 0.4);
            RestHealingScalar = (float)(0.6 + rng.NextDouble() * 0.8);

            // Hero
            HeroHealthScalar = (float)(0.7 + rng.NextDouble() * 0.6);
            HeroStartGoldScalar = (float)(0.8 + rng.NextDouble() * 0.4);
            
            int manaRoll = rng.Next(100);
            if (manaRoll < 5) HeroManaOffset = -1;
            else if (manaRoll > 95) HeroManaOffset = 1;
            else HeroManaOffset = 0;
        }

        public HierarchicalGenome Clone()
        {
            var clone = new HierarchicalGenome
            {
                GlobalDamageMultiplier = this.GlobalDamageMultiplier,
                GlobalHealthMultiplier = this.GlobalHealthMultiplier,
                GlobalBlockMultiplier = this.GlobalBlockMultiplier,
                GlobalManaCostMultiplier = this.GlobalManaCostMultiplier,
                GlobalGoldMultiplier = this.GlobalGoldMultiplier,
                
                EarlyGameDamageScaling = this.EarlyGameDamageScaling,
                MidGameDamageScaling = this.MidGameDamageScaling,
                LateGameDamageScaling = this.LateGameDamageScaling,
                EarlyGameHealthScaling = this.EarlyGameHealthScaling,
                MidGameHealthScaling = this.MidGameHealthScaling,
                LateGameHealthScaling = this.LateGameHealthScaling,
                EarlyGameBlockScaling = this.EarlyGameBlockScaling,
                MidGameBlockScaling = this.MidGameBlockScaling,
                LateGameBlockScaling = this.LateGameBlockScaling,
                
                CardTypeScalars = new Dictionary<CardType, float>(this.CardTypeScalars),
                CardStarScalars = new Dictionary<int, float>(this.CardStarScalars),
                EnemyStarScalars = new Dictionary<int, float>(this.EnemyStarScalars),
                
                RoomTypeWeights = new Dictionary<RoomType, float>(this.RoomTypeWeights),
                MonsterStarRatio = this.MonsterStarRatio,
                EliteStarRatio = this.EliteStarRatio,
                RestHealingScalar = this.RestHealingScalar,
                
                HeroHealthScalar = this.HeroHealthScalar,
                HeroStartGoldScalar = this.HeroStartGoldScalar,
                HeroManaOffset = this.HeroManaOffset,
                
                CardDamageOverrides = new Dictionary<string, float>(this.CardDamageOverrides),
                CardManaCostOverrides = new Dictionary<string, float>(this.CardManaCostOverrides),
                EnemyHealthOverrides = new Dictionary<string, float>(this.EnemyHealthOverrides)
            };
            return clone;
        }

        // SCALING CALCULATION METHODS
        
        public float GetProgressionScalar(int floor, ScalingType type)
        {
            float early, mid, late;
            
            switch (type)
            {
                case ScalingType.Damage:
                    early = EarlyGameDamageScaling;
                    mid = MidGameDamageScaling;
                    late = LateGameDamageScaling;
                    break;
                case ScalingType.Health:
                    early = EarlyGameHealthScaling;
                    mid = MidGameHealthScaling;
                    late = LateGameHealthScaling;
                    break;
                case ScalingType.Block:
                    early = EarlyGameBlockScaling;
                    mid = MidGameBlockScaling;
                    late = LateGameBlockScaling;
                    break;
                default:
                    return 1.0f;
            }
            
            if (floor <= 5) return early;
            if (floor <= 10) return mid;
            return late;
        }
        
        public int GetScaledCardDamage(CardData card, int floor)
        {
            float scalar = GlobalDamageMultiplier;
            scalar *= CardTypeScalars[card.Type];
            scalar *= CardStarScalars[card.StarRating];
            scalar *= GetProgressionScalar(floor, ScalingType.Damage);
            
            if (CardDamageOverrides.TryGetValue(card.Id, out float ovr))
                scalar *= ovr;
            
            var damageAction = card.Actions.FirstOrDefault(a => a.Type == ActionType.DealDamage);
            if (damageAction == null) return 0;
            
            return Math.Max(1, (int)Math.Round(damageAction.Value * scalar));
        }
        
        public int GetScaledCardBlock(CardData card, int floor)
        {
            float scalar = GlobalBlockMultiplier;
            scalar *= CardTypeScalars[card.Type];
            scalar *= CardStarScalars[card.StarRating];
            scalar *= GetProgressionScalar(floor, ScalingType.Block);
            
            var blockAction = card.Actions.FirstOrDefault(a => a.Type == ActionType.GainBlock);
            if (blockAction == null) return 0;
            
            return Math.Max(1, (int)Math.Round(blockAction.Value * scalar));
        }
        
        public int GetScaledCardCost(CardData card)
        {
            float scalar = GlobalManaCostMultiplier;
            
            if (CardManaCostOverrides.TryGetValue(card.Id, out float ovr))
                scalar *= ovr;
            
            int newCost = (int)Math.Round(card.ManaCost * scalar);
            return Math.Max(0, newCost);
        }
        
        public int GetScaledEnemyHealth(EnemyData enemy, int floor)
        {
            float scalar = GlobalHealthMultiplier;
            scalar *= EnemyStarScalars[enemy.StarRating];
            scalar *= GetProgressionScalar(floor, ScalingType.Health);
            
            if (EnemyHealthOverrides.TryGetValue(enemy.Id, out float ovr))
                scalar *= ovr;
            
            return Math.Max(1, (int)Math.Round(enemy.StartingHealth * scalar));
        }

        public enum ScalingType
        {
            Damage,
            Health,
            Block
        }
    }
}

File: python_roguelike/optimization/hierarchical_genome.py
"""
hierarchical_genome.py  –  Python port of HierarchicalGenome.cs

A hierarchical genome that reduces the parameter space from ~500 to ~50
parameters by using multiplicative layers and only overriding specific outliers.
"""

import copy
import math
import random
from dataclasses import dataclass, field
from enum import Enum
from typing import Dict


class ScalingType(Enum):
    Damage = "Damage"
    Health = "Health"
    Block = "Block"


CARD_TYPE_ATTACK = "Attack"
CARD_TYPE_SKILL = "Skill"
CARD_TYPE_POWER = "Power"


@dataclass
class HierarchicalGenome:
    global_damage_multiplier: float = 1.0
    global_health_multiplier: float = 1.0
    global_block_multiplier: float = 1.0
    global_mana_cost_multiplier: float = 1.0
    global_gold_multiplier: float = 1.0

    early_game_damage_scaling: float = 1.0
    mid_game_damage_scaling: float = 1.0
    late_game_damage_scaling: float = 1.0

    early_game_health_scaling: float = 1.0
    mid_game_health_scaling: float = 1.0
    late_game_health_scaling: float = 1.0

    early_game_block_scaling: float = 1.0
    mid_game_block_scaling: float = 1.0
    late_game_block_scaling: float = 1.0

    card_type_scalars: Dict[str, float] = field(default_factory=lambda: {
        CARD_TYPE_ATTACK: 1.0,
        CARD_TYPE_SKILL: 1.0,
        CARD_TYPE_POWER: 1.0,
    })

    card_star_scalars: Dict[int, float] = field(default_factory=lambda: {
        1: 1.0, 2: 1.0, 3: 1.0, 4: 1.0, 5: 1.0,
    })

    enemy_star_scalars: Dict[int, float] = field(default_factory=lambda: {
        1: 1.0, 2: 1.0, 3: 1.0, 4: 1.0, 5: 1.0,
    })

    room_type_weights: Dict[str, float] = field(default_factory=lambda: {
        "Monster": 45.0,
        "Elite": 12.0,
        "Event": 22.0,
        "Shop": 8.0,
        "Rest": 13.0,
    })

    monster_star_ratio: float = 0.5
    elite_star_ratio: float = 0.5
    rest_healing_scalar: float = 1.0

    hero_health_scalar: float = 1.0
    hero_start_gold_scalar: float = 1.0
    hero_mana_offset: int = 0

    difficulty_progression_rate: float = 0.03

    card_damage_overrides: Dict[str, float] = field(default_factory=dict)
    card_mana_cost_overrides: Dict[str, float] = field(default_factory=dict)
    enemy_health_overrides: Dict[str, float] = field(default_factory=dict)

    def randomize(self, rng: random.Random) -> None:
        self.global_damage_multiplier = 0.7 + rng.random() * 0.6
        self.global_health_multiplier = 0.7 + rng.random() * 0.6
        self.global_block_multiplier = 0.7 + rng.random() * 0.6
        self.global_mana_cost_multiplier = 0.85 + rng.random() * 0.3
        self.global_gold_multiplier = 0.7 + rng.random() * 0.6

        self.early_game_damage_scaling = 0.8 + rng.random() * 0.3
        self.mid_game_damage_scaling = 1.0 + rng.random() * 0.3
        self.late_game_damage_scaling = 1.2 + rng.random() * 0.3

        self.early_game_health_scaling = 0.8 + rng.random() * 0.3
        self.mid_game_health_scaling = 1.0 + rng.random() * 0.3
        self.late_game_health_scaling = 1.2 + rng.random() * 0.3

        self.early_game_block_scaling = 0.9 + rng.random() * 0.2
        self.mid_game_block_scaling = 1.0 + rng.random() * 0.2
        self.late_game_block_scaling = 1.0 + rng.random() * 0.2

        for key in list(self.card_type_scalars):
            self.card_type_scalars[key] = 0.85 + rng.random() * 0.3

        for key in list(self.card_star_scalars):
            self.card_star_scalars[key] = 0.9 + rng.random() * 0.2

        for key in list(self.enemy_star_scalars):
            self.enemy_star_scalars[key] = 0.9 + rng.random() * 0.2

        self.room_type_weights["Monster"] = 30 + rng.random() * 30
        self.room_type_weights["Elite"] = 5 + rng.random() * 20
        self.room_type_weights["Event"] = 10 + rng.random() * 25
        self.room_type_weights["Shop"] = 5 + rng.random() * 15
        self.room_type_weights["Rest"] = 5 + rng.random() * 20

        self.monster_star_ratio = 0.3 + rng.random() * 0.4
        self.elite_star_ratio = 0.3 + rng.random() * 0.4
        self.rest_healing_scalar = 0.6 + rng.random() * 0.8

        self.hero_health_scalar = 0.7 + rng.random() * 0.6
        self.hero_start_gold_scalar = 0.8 + rng.random() * 0.4
        self.difficulty_progression_rate = rng.random() * 0.06

        mana_roll = rng.randint(0, 99)
        if mana_roll < 5:
            self.hero_mana_offset = -1
        elif mana_roll > 95:
            self.hero_mana_offset = 1
        else:
            self.hero_mana_offset = 0

    def clone(self) -> "HierarchicalGenome":
        return copy.deepcopy(self)


    def get_progression_scalar(self, floor: int, scaling_type: ScalingType) -> float:
        if scaling_type == ScalingType.Damage:
            early, mid, late = (self.early_game_damage_scaling,
                                self.mid_game_damage_scaling,
                                self.late_game_damage_scaling)
        elif scaling_type == ScalingType.Health:
            early, mid, late = (self.early_game_health_scaling,
                                self.mid_game_health_scaling,
                                self.late_game_health_scaling)
        elif scaling_type == ScalingType.Block:
            early, mid, late = (self.early_game_block_scaling,
                                self.mid_game_block_scaling,
                                self.late_game_block_scaling)
        else:
            return 1.0

        if floor <= 5:
            return early
        if floor <= 10:
            return mid
        return late

    def get_floor_difficulty_multiplier(self, floor: int) -> float:
        t = (floor - 1) / 14.0
        s_curve = 1.0 / (1.0 + math.exp(-10 * (t - 0.5)))
        return 1.0 + self.difficulty_progression_rate * 5.0 * s_curve


    PARAM_SPECS = [
        ("global_damage_multiplier", 0.7, 1.3),
        ("global_health_multiplier", 0.7, 1.3),
        ("global_block_multiplier", 0.7, 1.3),
        ("global_mana_cost_multiplier", 0.85, 1.15),
        ("global_gold_multiplier", 0.7, 1.3),
        ("early_game_damage_scaling", 0.7, 1.2),
        ("mid_game_damage_scaling", 0.9, 1.4),
        ("late_game_damage_scaling", 1.1, 1.6),
        ("early_game_health_scaling", 0.7, 1.2),
        ("mid_game_health_scaling", 0.9, 1.4),
        ("late_game_health_scaling", 1.1, 1.6),
        ("early_game_block_scaling", 0.8, 1.1),
        ("mid_game_block_scaling", 0.9, 1.2),
        ("late_game_block_scaling", 0.9, 1.2),
        ("card_type_Attack", 0.85, 1.15),
        ("card_type_Skill", 0.85, 1.15),
        ("card_type_Power", 0.85, 1.15),
        ("card_star_1", 0.9, 1.1),
        ("card_star_2", 0.9, 1.1),
        ("card_star_3", 0.9, 1.1),
        ("card_star_4", 0.9, 1.1),
        ("card_star_5", 0.9, 1.1),
        ("enemy_star_1", 0.9, 1.1),
        ("enemy_star_2", 0.9, 1.1),
        ("enemy_star_3", 0.9, 1.1),
        ("enemy_star_4", 0.9, 1.1),
        ("enemy_star_5", 0.9, 1.1),
        ("room_Monster", 30.0, 60.0),
        ("room_Elite", 5.0, 25.0),
        ("room_Event", 10.0, 35.0),
        ("room_Shop", 5.0, 20.0),
        ("room_Rest", 5.0, 25.0),
        ("monster_star_ratio", 0.2, 0.8),
        ("elite_star_ratio", 0.2, 0.8),
        ("rest_healing_scalar", 0.5, 1.5),
        ("hero_health_scalar", 0.7, 1.3),
        ("hero_start_gold_scalar", 0.8, 1.2),
        ("difficulty_progression_rate", 0.0, 0.06),
    ]

    N_CONTINUOUS = len(PARAM_SPECS)

    @classmethod
    def get_bounds(cls):
        """Returns (xl, xu) arrays for pymoo."""
        import numpy as np
        xl = np.array([s[1] for s in cls.PARAM_SPECS], dtype=np.float64)
        xu = np.array([s[2] for s in cls.PARAM_SPECS], dtype=np.float64)
        return xl, xu

    def to_vector(self):
        """Flatten genome into a numpy array for pymoo."""
        import numpy as np
        x = np.zeros(self.N_CONTINUOUS, dtype=np.float64)
        i = 0
        x[i] = self.global_damage_multiplier; i += 1
        x[i] = self.global_health_multiplier; i += 1
        x[i] = self.global_block_multiplier; i += 1
        x[i] = self.global_mana_cost_multiplier; i += 1
        x[i] = self.global_gold_multiplier; i += 1
        x[i] = self.early_game_damage_scaling; i += 1
        x[i] = self.mid_game_damage_scaling; i += 1
        x[i] = self.late_game_damage_scaling; i += 1
        x[i] = self.early_game_health_scaling; i += 1
        x[i] = self.mid_game_health_scaling; i += 1
        x[i] = self.late_game_health_scaling; i += 1
        x[i] = self.early_game_block_scaling; i += 1
        x[i] = self.mid_game_block_scaling; i += 1
        x[i] = self.late_game_block_scaling; i += 1
        x[i] = self.card_type_scalars[CARD_TYPE_ATTACK]; i += 1
        x[i] = self.card_type_scalars[CARD_TYPE_SKILL]; i += 1
        x[i] = self.card_type_scalars[CARD_TYPE_POWER]; i += 1
        for s in range(1, 6):
            x[i] = self.card_star_scalars[s]; i += 1
        for s in range(1, 6):
            x[i] = self.enemy_star_scalars[s]; i += 1
        for rt in ["Monster", "Elite", "Event", "Shop", "Rest"]:
            x[i] = self.room_type_weights[rt]; i += 1
        x[i] = self.monster_star_ratio; i += 1
        x[i] = self.elite_star_ratio; i += 1
        x[i] = self.rest_healing_scalar; i += 1
        x[i] = self.hero_health_scalar; i += 1
        x[i] = self.hero_start_gold_scalar; i += 1
        x[i] = self.difficulty_progression_rate; i += 1
        return x

    @classmethod
    def from_vector(cls, x) -> "HierarchicalGenome":
        """Reconstruct genome from a numpy vector (pymoo decision variable)."""
        g = cls()
        i = 0
        g.global_damage_multiplier = float(x[i]); i += 1
        g.global_health_multiplier = float(x[i]); i += 1
        g.global_block_multiplier = float(x[i]); i += 1
        g.global_mana_cost_multiplier = float(x[i]); i += 1
        g.global_gold_multiplier = float(x[i]); i += 1

        g.early_game_damage_scaling = float(x[i]); i += 1
        g.mid_game_damage_scaling = float(x[i]); i += 1
        g.late_game_damage_scaling = float(x[i]); i += 1
        g.early_game_health_scaling = float(x[i]); i += 1
        g.mid_game_health_scaling = float(x[i]); i += 1
        g.late_game_health_scaling = float(x[i]); i += 1
        g.early_game_block_scaling = float(x[i]); i += 1
        g.mid_game_block_scaling = float(x[i]); i += 1
        g.late_game_block_scaling = float(x[i]); i += 1

        g.card_type_scalars[CARD_TYPE_ATTACK] = float(x[i]); i += 1
        g.card_type_scalars[CARD_TYPE_SKILL] = float(x[i]); i += 1
        g.card_type_scalars[CARD_TYPE_POWER] = float(x[i]); i += 1

        for s in range(1, 6):
            g.card_star_scalars[s] = float(x[i]); i += 1
        for s in range(1, 6):
            g.enemy_star_scalars[s] = float(x[i]); i += 1

        for rt in ["Monster", "Elite", "Event", "Shop", "Rest"]:
            g.room_type_weights[rt] = float(x[i]); i += 1

        g.monster_star_ratio = float(x[i]); i += 1
        g.elite_star_ratio = float(x[i]); i += 1
        g.rest_healing_scalar = float(x[i]); i += 1

        g.hero_health_scalar = float(x[i]); i += 1
        g.hero_start_gold_scalar = float(x[i]); i += 1
        g.difficulty_progression_rate = float(x[i]); i += 1
        return g
